ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -162,18 +307,52 @@ public boolean equals(Object o) {
     if (!version.equals(blobId.version)) {
       return false;
     }
-    if (!partitionId.equals(blobId.partitionId)) {
+    if (!flag.equals(blobId.flag)) {
       return false;
     }
-    if (!uuid.equals(blobId.uuid)) {
+    if (!datacenterId.equals(blobId.datacenterId)) {
       return false;
     }
-
-    return true;
+    if (!accountId.equals(blobId.accountId)) {
+      return false;
+    }
+    if (!containerId.equals(blobId.containerId)) {
+      return false;
+    }
+    if (!partitionId.equals(blobId.partitionId)) {
+      return false;
+    }
+    return uuid.equals(blobId.uuid);
   }
 
   @Override
   public int hashCode() {
     return Utils.hashcode(new Object[]{version, partitionId, uuid});
   }
+
+  /**
+   * Populate the missing fields with legacy values, and determine the version of format to serialize. Only
+   * when all the four fields of {@code flag}, {@code datacenterId}, {@code accountId}, and {@code containerId}
+   * are available, the serialization format version will be set to {@link #BLOB_ID_V2}, otherwise it will be
+   * {@link #BLOB_ID_V1}.
+   */
+  private void populateMissingFieldsAndDetermineVersion() {
+    version = flag == null && accountId == null && containerId == null && datacenterId == null ? BlobId.BLOB_ID_V1
```

Comment:
So here is the thing. This change will have to go in the same way as other version changes. All readers should know how to deserialize a V2 BlobId. You will also have to ensure that once you switch to V2 a node does not go back to V1. We cannot have V1 and V2 blobs go in at the same time as the store will keep creating new index segments due to the size change, which is not good.

Please keep that in mind when making this change. It might mean that you keep a "Current_BlobId_Version" which is set to V1 and at some point you switch to V2 and at that point even if these fields are null you will populate it with -1. 

I am wondering how we should deal with this during deployment (in the interim period when one frontend writes in V1 and another in V2 and they both put to the same server). 

Let us think this through and flesh this out properly.